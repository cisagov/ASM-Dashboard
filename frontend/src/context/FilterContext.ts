import React, { useContext, createContext } from 'react';

export interface OrganizationShallow {
  regionId: string;
  name: string;
  id: string;
  rootDomains: string[];
}

export type Region = string;

export interface FilterContext {
  regions: Array<Region>;
  organizations: OrganizationShallow[];
  setRegions: (regions: FilterContext['regions']) => void;
  setOrganizations: (organizations: FilterContext['organizations']) => void;
  addOrganization: (organization: OrganizationShallow) => void;
  removeOrganization: (organization: OrganizationShallow) => void;
}

export const FilterContext = React.createContext<FilterContext>(undefined!);

export const useFilterContext = (): FilterContext => {
  return useContext(FilterContext);
};
