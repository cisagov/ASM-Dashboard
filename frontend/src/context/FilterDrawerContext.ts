import React, { useContext } from 'react';

export interface FilterDrawerContextType {
  isFilterDrawerOpen: boolean;
  setIsFilterDrawerOpen: (isFilterDrawerOpen: boolean) => void;
}

export const FilterDrawerContext = React.createContext<FilterDrawerContextType>(
  undefined!
);

export const useFilterDrawerContext = (): FilterDrawerContextType => {
  return useContext(FilterDrawerContext);
};
