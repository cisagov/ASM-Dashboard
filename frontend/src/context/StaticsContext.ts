import React, { useContext } from 'react';

export type Region = string;

export interface StaticsContextType {
  regions: string[];
  setRegions: (regions: StaticsContextType['regions']) => void;
}

export const StaticsContext = React.createContext<StaticsContextType>(
  undefined!
);

export const useStaticsContext = (): StaticsContextType => {
  return useContext(StaticsContext);
};
