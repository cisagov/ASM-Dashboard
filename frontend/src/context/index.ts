export * from './AuthContext';
export * from './AuthContextProvider';
export * from './Theme';
export * from './SearchProvider';
export * from './SavedSearchContext';
export * from './SavedSearchContextProvider';
export * from './FilterDrawerContext';
export * from './FilterDrawerContextProvider';
