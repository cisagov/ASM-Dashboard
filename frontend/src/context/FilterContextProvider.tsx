import React, { useCallback, useState } from 'react';
import { FilterContext, OrganizationShallow, Region } from './FilterContext';
import { usePersistentState } from 'hooks';

interface FilterContextProviderProps {
  children: React.ReactNode;
}

export const FilterContextProvider: React.FC<FilterContextProviderProps> = ({
  children
}) => {
  const [regions, setRegions] = usePersistentState<Region[]>(
    'filter-regions',
    []
  );

  const [organizations, setOrganizations] = usePersistentState<
    OrganizationShallow[]
  >('filter-organizations', []);

  const addOrganization = useCallback(
    (organization: OrganizationShallow) => {
      const newOrgs = [...organizations, organization];
      setOrganizations(newOrgs);
    },
    [organizations]
  );

  const removeOrganization = useCallback(
    (organization: OrganizationShallow) => {
      setOrganizations(
        organizations.filter((org) => org.id !== organization.id)
      );
    },
    [organizations]
  );

  return (
    <FilterContext.Provider
      value={{
        regions,
        organizations,
        setOrganizations,
        setRegions,
        removeOrganization,
        addOrganization
      }}
    >
      {children}
    </FilterContext.Provider>
  );
};
