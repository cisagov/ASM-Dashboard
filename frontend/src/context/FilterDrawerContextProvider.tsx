import React from 'react';
import { FilterDrawerContext } from 'context/FilterDrawerContext';
import { useFilterDrawerContext } from 'context/FilterDrawerContext';
import { usePersistentState } from 'hooks';

interface FilterDrawerContextProviderProps {
  children: React.ReactNode;
}

export const FilterDrawerContextProvider: React.FC<
  FilterDrawerContextProviderProps
> = ({ children }) => {
  const [isFilterDrawerOpen, setIsFilterDrawerOpen] = usePersistentState(
    'filterDrawerOpen',
    false
  );

  return (
    <FilterDrawerContext.Provider
      value={{ isFilterDrawerOpen, setIsFilterDrawerOpen }}
    >
      {children}
    </FilterDrawerContext.Provider>
  );
};
