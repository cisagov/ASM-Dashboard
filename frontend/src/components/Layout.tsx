import React, {
  PropsWithChildren,
  useCallback,
  useEffect,
  useMemo,
  useState
} from 'react';
import { styled } from '@mui/material/styles';
import { useLocation } from 'react-router-dom';
import { Box, Drawer, ScopedCssBaseline, useMediaQuery } from '@mui/material';
import { Header, GovBanner } from 'components';
import { useUserActivityTimeout } from 'hooks/useUserActivityTimeout';
import { useAuthContext } from 'context/AuthContext';
import UserInactiveModal from './UserInactivityModal/UserInactivityModal';
import { CrossfeedFooter } from './Footer';
import { RSCFooter } from './ReadySetCyber/RSCFooter';
import { RSCHeader } from './ReadySetCyber/RSCHeader';
import { SkipToMainContent } from './SkipToMainContent/index';
import { matchPath } from 'utils/matchPath';
import { drawerWidth, FilterDrawerV2 } from './FilterDrawerV2';
import { usePersistentState } from 'hooks';
import { useTheme } from '@mui/system';
import { withSearch } from '@elastic/react-search-ui';
import { ContextType } from 'context';
import { useUserTypeFilters } from 'hooks/useUserTypeFilters';
import { useStaticsContext } from 'context/StaticsContext';
interface LayoutProps {
  children: React.ReactNode;
}

const GLOBAL_ADMIN = 3;
const REGIONAL_ADMIN = 2;
const STANDARD_USER = 1;

const Main = styled('main', { shouldForwardProp: (prop) => prop !== 'open' })<{
  open?: boolean;
}>(({ theme, open }) => ({
  flexGrow: 1,
  height: 'calc(100vh - 24px)',
  maxHeight: 'calc(100vh - 24px)',
  overflow: 'scroll',
  '&::-webkit-scrollbar': {
    display: 'none'
  },
  transition: theme.transitions.create('margin', {
    easing: theme.transitions.easing.sharp,
    duration: theme.transitions.duration.leavingScreen
  }),
  [theme.breakpoints.up('lg')]: {
    marginLeft: `-${drawerWidth}px`
  },
  [theme.breakpoints.down('lg')]: {
    marginLeft: 0
  },
  ...(open && {
    transition: theme.transitions.create('margin', {
      easing: theme.transitions.easing.easeOut,
      duration: theme.transitions.duration.enteringScreen
    }),
    [theme.breakpoints.up('lg')]: {
      marginLeft: 0
    },
    [theme.breakpoints.down('lg')]: {
      marginLeft: 0
    }
  })
}));

export const Layout: React.FC<PropsWithChildren<ContextType>> = ({
  children,
  filters,
  addFilter,
  removeFilter
}) => {
  const { logout, user } = useAuthContext();

  useEffect(() => {
    localStorage.setItem('es-search-filters', JSON.stringify(filters));
  }, [filters]);

  const { regions } = useStaticsContext();

  const theme = useTheme();
  const [isFilterDrawerOpen, setIsFilterDrawerOpen] = usePersistentState(
    'isFilterDrawerOpen',
    false
  );

  const userLevel = useMemo(() => {
    if (user && user.isRegistered) {
      if (user.userType === 'standard') {
        return STANDARD_USER;
      } else if (user.userType === 'globalAdmin') {
        return GLOBAL_ADMIN;
      } else if (
        user.userType === 'regionalAdmin' ||
        user.userType === 'globalView'
      ) {
        return REGIONAL_ADMIN;
      }
      return 0;
    }
    return 0;
  }, [user]);

  const [loggedIn, setLoggedIn] = useState<boolean>(
    user !== null && user !== undefined ? true : false
  );
  const { isTimedOut, resetTimeout } = useUserActivityTimeout(
    14 * 60 * 1000, // set to 14 minutes of inactivity to notify user
    loggedIn
  );

  const handleCountdownEnd = useCallback(
    (shouldLogout: boolean) => {
      if (shouldLogout) {
        logout();
      } else {
        resetTimeout();
      }
    },
    [logout, resetTimeout]
  );

  const { pathname } = useLocation();

  useEffect(() => {
    const pathsAllowed = ['/', '/inventory'];
    if (!matchPath(pathsAllowed, pathname)) {
      setIsFilterDrawerOpen(false);
    }
  }, [pathname]);

  useEffect(() => {
    // set logged in if use exists then set true, otherwise set false
    if (user) setLoggedIn(true);
    else setLoggedIn(false);
  }, [user]);

  const initialFiltersForUser = useUserTypeFilters(
    filters,
    addFilter,
    removeFilter,
    regions,
    user,
    userLevel
  );

  useEffect(() => {
    initialFiltersForUser.forEach((filter) => {
      filter.values.forEach((val) => {
        addFilter(filter.field, val, filter.type);
      });
    });
  }, [regions, user]);

  const isMobile = useMediaQuery(theme.breakpoints.down('lg'));

  return (
    <StyledScopedCssBaseline classes={{ root: classes.overrides }}>
      <div className={classes.root}>
        <UserInactiveModal
          isOpen={isTimedOut}
          onCountdownEnd={handleCountdownEnd}
          countdown={60} // 60 second timer for user inactivity timeout
        />
        <div style={{ display: 'flex' }}>
          <GovBanner />
          <SkipToMainContent />
        </div>
        {!pathname.includes('/readysetcyber') ? (
          <>
            <div
              style={{
                display: 'flex',
                flexDirection: 'row',
                height: '100vh'
              }}
            >
              {userLevel > 0 &&
              matchPath(
                [
                  '/',
                  '/inventory',
                  '/inventory/domains',
                  '/inventory/vulnerabilities'
                ],
                pathname
              ) ? (
                <FilterDrawerV2
                  setIsFilterDrawerOpen={setIsFilterDrawerOpen}
                  isFilterDrawerOpen={isFilterDrawerOpen}
                  isMobile={isMobile}
                />
              ) : (
                <Drawer sx={{ width: drawerWidth }} variant="persistent">
                  <Box sx={{ width: drawerWidth }} />
                </Drawer>
              )}
              <Main open={isFilterDrawerOpen}>
                <Header
                  isFilterDrawerOpen={isFilterDrawerOpen}
                  setIsFilterDrawerOpen={setIsFilterDrawerOpen}
                />

                <Box
                  display="block"
                  position="relative"
                  flex="1"
                  height="calc(100vh - 64px - 72px - 24px)"
                  overflow="scroll"
                  sx={{
                    '&::-webkit-scrollbar': {
                      display: 'none'
                    }
                  }}
                  zIndex={16}
                >
                  {children}
                </Box>

                {/* <div className="main-content" id="main-content" tabIndex={-1} />
                {pathname === '/inventory' ? (
                  children
                ) : (
                  <div className={classes.content}>{children}</div>
                )} */}

                <CrossfeedFooter />
              </Main>
            </div>
          </>
        ) : (
          <>
            <RSCHeader />
            <div className={classes.content}>{children}</div>
            <RSCFooter />
          </>
        )}
      </div>
    </StyledScopedCssBaseline>
  );
};

export const LayoutWithSearch = withSearch(
  ({ addFilter, removeFilter, filters }: ContextType) => ({
    addFilter,
    removeFilter,
    filters
  })
)(Layout);

//Styling
const PREFIX = 'Layout';

const classes = {
  root: `${PREFIX}-root`,
  overrides: `${PREFIX}-overrides`,
  content: `${PREFIX}-content`
};

const StyledScopedCssBaseline = styled(ScopedCssBaseline)(({ theme }) => ({
  [`& .${classes.root}`]: {
    position: 'relative',
    height: '100vh',
    display: 'flex',
    flexFlow: 'column nowrap'
    // overflow: 'auto'
  },

  [`& .${classes.overrides}`]: {
    WebkitFontSmoothing: 'unset',
    MozOsxFontSmoothing: 'unset'
  },

  [`& .${classes.content}`]: {
    flex: '1',
    display: 'block',
    position: 'relative',
    height: 'calc(100vh - 24px)',
    overflow: 'scroll'
  }
}));
