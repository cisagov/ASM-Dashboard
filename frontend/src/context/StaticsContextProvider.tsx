import React, { useCallback, useEffect, useMemo } from 'react';

import { usePersistentState } from 'hooks';
import { StaticsContext } from './StaticsContext';
import { useAuthContext } from './AuthContext';

interface StaticsContextProviderProps {
  children: React.ReactNode;
}

export const StaticsContextProvider: React.FC<StaticsContextProviderProps> = ({
  children
}) => {
  const [regions, setRegions] = usePersistentState<string[]>(
    'filter-regions',
    []
  );

  const { apiGet } = useAuthContext();
  const fetchRegions = useCallback(async () => {
    try {
      const results = await apiGet('/regions');
      setRegions(
        results
          .sort((a: string, b: string) => parseInt(a) - parseInt(b))
          .map((region: { regionId: string }) => region.regionId)
      );
    } catch (e) {
      console.log(e);
    }
  }, []);

  useEffect(() => {
    fetchRegions();
  }, []);

  const memoizedRegions = useMemo(() => {
    return regions;
  }, [regions]);

  return (
    <StaticsContext.Provider
      value={{
        regions: memoizedRegions,
        setRegions
      }}
    >
      {children}
    </StaticsContext.Provider>
  );
};
