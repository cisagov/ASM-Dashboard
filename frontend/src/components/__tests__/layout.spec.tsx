import React from 'react';
import { render } from 'test-utils/test-utils';
import { Layout } from '../Layout';
import { StaticsContext, StaticsContextType } from 'context/StaticsContext';
import { ContextType, SearchProvider } from 'context';

const testContext: ContextType = {
  addFilter: (field: string, value: any, type: 'any' | 'all' | 'none') => {},
  removeFilter: (field: string, value: any, type: 'any' | 'all' | 'none') => {},
  filters: [],
  clearFilters: () => {},
  saveSearch: '',
  setSearchTerm: (s: string, opts?: any) => {},
  autocompletedResults: [],
  autocompletedResultsRequestId: 'string',
  autocompletedSuggestions: '',
  current: 1,
  error: 'string',
  facets: '',
  isLoading: false,
  pagingEnd: 1,
  pagingStart: 1,
  requestId: 'string',
  reset: () => {},
  resultSearchTerm: 'string',
  results: [],
  resultsPerPage: 1,
  searchTerm: 'string',
  setCurrent: (current: number) => {},
  setFilter: () => {},
  setResultsPerPage: () => {},
  setSort: (field: 'string', direction: 'asc' | 'desc') => {},
  sortDirection: '',
  sortField: 'string',
  totalPages: 1,
  totalResults: 1,
  wasSearched: false,
  noResults: false
};

const value: StaticsContextType = {
  regions: [],
  setRegions: (regions: string[]) => {}
};

jest.mock('components/Header', () => ({
  Header: () => <div>HEADER</div>
}));
jest.mock('components/GovBanner', () => ({
  GovBanner: () => <div>GOV_BANNER</div>
}));

afterAll(() => {
  jest.restoreAllMocks();
});

describe('Layout component', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(
      <SearchProvider>
        <StaticsContext.Provider value={value}>
          <Layout {...testContext} />
        </StaticsContext.Provider>
      </SearchProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders children', () => {
    const { getByText } = render(
      <SearchProvider>
        <StaticsContext.Provider value={value}>
          <Layout {...testContext}>some children</Layout>
        </StaticsContext.Provider>
      </SearchProvider>
    );
    expect(getByText('some children')).toBeInTheDocument();
  });
});
