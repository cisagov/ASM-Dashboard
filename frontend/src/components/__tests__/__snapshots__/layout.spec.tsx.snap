// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiScopedCssBaseline-root Layout-overrides css-dwi7y5-MuiScopedCssBaseline-root"
  >
    <div
      class="Layout-root"
    >
      <div
        aria-label="warning label"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardWarning MuiAlert-standard css-to2f46-MuiPaper-root-MuiAlert-root"
        role="alert"
      >
        <div
          class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
            data-testid="ReportProblemOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 5.99L19.53 19H4.47L12 5.99M12 2L1 21h22L12 2zm1 14h-2v2h2v-2zm0-6h-2v4h2v-4z"
            />
          </svg>
        </div>
        <div
          class="MuiAlert-message css-1pxa9xg-MuiAlert-message"
        >
          <div />
        </div>
      </div>
      <div>
        GOV_BANNER
      </div>
      <div>
        HEADER
      </div>
      <div
        class="Layout-content"
      />
      <div
        class="css-1ycjho2"
      >
        <div
          class="Footer-footerBox MuiBox-root css-0"
        >
          <div
            class="MuiGrid-root MuiGrid-container Footer-footerContainer css-11lq3yg-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-2 Footer-footerLogo css-6vomyv-MuiGrid-root"
            >
              <a
                class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-smdm7r-MuiTypography-root-MuiLink-root"
                href="/"
              >
                <img
                  alt="Crossfeed Icon Navigate Home"
                  src="crossfeed.svg"
                />
              </a>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-2 Footer-footerNavItem css-6vomyv-MuiGrid-root"
            >
              <a
                class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways Footer-footerNavLink css-smdm7r-MuiTypography-root-MuiLink-root"
                href="/"
              >
                Home
              </a>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-2 Footer-footerNavItem css-6vomyv-MuiGrid-root"
            >
              <p>
                <a
                  class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways Footer-footerNavLink css-smdm7r-MuiTypography-root-MuiLink-root"
                  href="https://docs.crossfeed.cyber.dhs.gov/"
                  target="_blank"
                >
                  Documentation
                </a>
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-2 Footer-footerNavItem css-6vomyv-MuiGrid-root"
            >
              <p>
                <a
                  class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways Footer-footerNavLink css-smdm7r-MuiTypography-root-MuiLink-root"
                  href="https://www.cisa.gov"
                  target="_blank"
                >
                  CISA Homepage
                </a>
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-2 Footer-footerNavItem css-6vomyv-MuiGrid-root"
            >
              <p>
                <a
                  class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways Footer-footerNavLink css-smdm7r-MuiTypography-root-MuiLink-root"
                  href="mailto:vulnerability@cisa.dhs.gov"
                >
                  Contact Us
                </a>
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-2 Footer-footerNavItem css-6vomyv-MuiGrid-root"
            >
              <p>
                <a
                  class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways Footer-footerNavLink css-smdm7r-MuiTypography-root-MuiLink-root"
                  href="/"
                >
                  Logout
                </a>
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
